crates/kornia-imgproc/src/calibration/find_contours.rs

```diff
@@ -0,0 +1,287 @@
+pub enum BorderType {
+    /// The outer perimeter of a foreground region.
+    Outer,
+    /// The perimeter of a background “hole” fully enclosed by foreground.
+    Hole,
+}
+
+/// A detected contour in the image.
+///
+/// - `points` is the ordered list of (x,y) pixel coordinates along the border.
+/// - `border_type` tells you if this is an `Outer` contour or a `Hole`.
+/// - `parent` is an optional index into the returned `Vec<Contour>` of the
+///   contour that immediately encloses this one (if any).
+#[derive(Debug, Clone)]
+pub struct Contour {
+    /// The (x,y) coordinates of each border pixel in tracing order.
+    pub points: Vec<(usize, usize)>,
+    /// Whether this contour is an outer border or a hole.
+    pub border_type: BorderType,
+    /// Index of the parent contour (if this is a hole inside another contour).
+    pub parent: Option<usize>,
+}
+
+impl Contour {
+    /// Create a new `Contour`.
+    ///
+    /// # Arguments
+    ///
+    /// * `points` — the sequence of (x,y) coordinates along the border.
+    /// * `border_type` — `BorderType::Outer` for an outer perimeter, or
+    ///   `BorderType::Hole` for an enclosed hole.
+    /// * `parent` — `Some(idx)` if this contour is enclosed by the contour at
+    ///   `contours[idx]`, or `None` if it has no enclosing parent.
+    pub fn new(
+        points: Vec<(usize, usize)>,
+        border_type: BorderType,
+        parent: Option<usize>,
+    ) -> Self {
+        Self {
+            points,
+            border_type,
+            parent,
+        }
+    }
+}
+
+/// Find all contours in a grayscale image by thresholding and applying
+/// the Suzuki–Abe border‐following algorithm.
+///
+/// # Arguments
+///
+/// * `image` — input grayscale image (`Image<u8,1>`).
+/// * `threshold` — any pixel value > this is treated as “foreground”.
+///
+/// # Returns
+///
+/// A `Vec<Contour>` listing every outer and hole border found in scan order.
+pub fn find_contours(image: &Image<u8, 1>, threshold: u8) -> Vec<Contour> {
+    let w = image.width() as usize;
+    let h = image.height() as usize;
+    // binary mask: 1=fg, 0=bg
+    let mut bin = vec![0u8; w * h];
+    // labels: 0=unvisited, >1=contour IDs
+    let mut labels = vec![0i32; w * h];
+    let idx = |x: usize, y: usize| y * w + x;
+
+    // converting to binary based on threshold
+    for y in 0..h {
+        for x in 0..w {
+            let v = *image.get_pixel(x, y, 0).unwrap();
+            bin[idx(x, y)] = if v > threshold { 1 } else { 0 };
+        }
+    }
+
+    // 8‐connected neighbor offsets in clockwise order: E, SE, S, SW, W, NW, N, NE
+    let nbrs: [(i32, i32); 8] = [
+        (1, 0),
+        (1, 1),
+        (0, 1),
+        (-1, 1),
+        (-1, 0),
+        (-1, -1),
+        (0, -1),
+        (1, -1),
+    ];
+
+    let mut contours = Vec::new();
+    let mut nbd = 2; // next contour ID
+
+    // raster scan
+    for y in 0..h {
+        for x in 0..w {
+            let i = idx(x, y);
```

Comment:
this can be split into a couple of calls in the parent loop to avoid computing y*w every time